lib/ddtrace/contrib/configuration/resolvers/pattern_resolver.rb

```diff
@@ -10,17 +10,23 @@ class PatternResolver < Datadog::Contrib::Configuration::Resolver
           def resolve(name)
             # Try to find a matching pattern
             matching_pattern = patterns.find do |pattern|
-              # Rubocop incorrectly thinks assignment is done here...
-              # rubocop:disable Style/ConditionalAssignment
               if pattern.is_a?(Proc)
                 pattern === name
+              elsif pattern.is_a?(Regexp)
```

Comment:
I recommend using `#===` for all the supported types, like we do in our [SimpleMatcher]( https://github.com/DataDog/dd-trace-rb/blob/dbcac580b0bbf314901ded18be7143eaaa5265cb/lib/ddtrace/sampling/matcher.rb#L31-L39). In this case, I don't think we should type check, but instead document that patterns must respond to `#===`.

You _might_ need to `#to_s` the `name`, but I'm not sure on that.

Reply:
So i've updated this whole pattern to the, far simpler

```
          def resolve(name)
            # Try to find a matching pattern
            matching_pattern = patterns.find do |pattern|
              if pattern.respond_to?(:===)
                (pattern === name) || (pattern === name.to_s) || (pattern == name)
              end
            end

            # Return match or default
            matching_pattern
          end
```

For the cases in `(pattern === name) || (pattern === name.to_s) || (pattern == name)`: 

- The first case is needed for when the `pattern` is a Proc or Regexp and the `name` is a string (or whatever the Proc or Regexp expect as input). As far as I can tell the `Proc` `===` [implementation is an alias for `.call`](https://github.com/ruby/ruby/blob/fa7484019d4321f8c0775a4811574dccd29bcaea/proc.c#L877) and Regexp `===` implementation is [equivalent to `regex =~ str >= 0`](https://blog.arkency.com/the-equals-equals-equals-case-equality-operator-in-ruby/)
- The second case is needed for when the `pattern` is a Regexp but the `name` is something that _coerces_ to a string, like `URL('<url>')`, which is part of our [pattern resolver spec](https://github.com/DataDog/dd-trace-rb/blob/817a0827d0c6edeb41b37d7ea8d321ce09e1547a/spec/ddtrace/contrib/configuration/resolvers/pattern_resolver_spec.rb#L21). 
- The final case is needed when the `pattern` is a regexp or proc and then `name` is also a Regexp or Proc, which is how `resolver.resolve` gets used within [the Configurable `add_configuration` method](https://github.com/DataDog/dd-trace-rb/blob/817a0827d0c6edeb41b37d7ea8d321ce09e1547a/lib/ddtrace/contrib/configurable.rb#L60-L64)

Lmk what you think

Reply:
Can't case 1 and 2 be handled only by 2?

Also, I don't see how in 3 we'd receive a proc or regex as the name parameter. I tried to follow the method you linked, but it seems like we are receiving a regular key there. Also, was this case supported before? I don't see a raw `==` in the original code.

Reply:
1 and 2 are in opposition unfortunately if we aren't going to differentiate between a Proc and Regexp

Option 1 covers the case where we don't want to coerce the `name` for procs, [this is covered in spec here for Procs](https://github.com/DataDog/dd-trace-rb/blob/master/spec/ddtrace/contrib/configuration/resolvers/pattern_resolver_spec.rb#L39-L43)

Option 2 covers the case where we _do_ want to coerce the `name` for regexp, [this is covered in spec here for Regexp](https://github.com/DataDog/dd-trace-rb/blob/master/spec/ddtrace/contrib/configuration/resolvers/pattern_resolver_spec.rb#L17-L24)

That being said I can update this to check explicitly for a `Proc` to decide between using option 1 or 2. Also can remove the explicit `respond_to?(:===)`.

The need for option 3 is because we use resolver.resolve with `key` inputs that are both 

- the actal host, url string or string coercablle that we want to match against a pattern, [like here](https://github.com/DataDog/dd-trace-rb/blob/817a0827d0c6edeb41b37d7ea8d321ce09e1547a/lib/ddtrace/contrib/ethon/easy_patch.rb#L132)

- but also the `:describes` block (potentially a proc or regex) itself, [seen here](https://github.com/DataDog/dd-trace-rb/blob/817a0827d0c6edeb41b37d7ea8d321ce09e1547a/lib/ddtrace/contrib/extensions.rb#L60).

There's a lot of room for improvement in the way we're using resolvers but for now to fit the extisting usage the pattern resolver has to be pretty permissive with what it can resolve against